Подберём к вопросу похожий на него вопрос.

In [2]:
!git clone https://github.com/vladislavneon/RuBQ.git

fatal: destination path 'RuBQ' already exists and is not an empty directory.


In [5]:
import json

filepath = "./RuBQ/RuBQ_2.0/RuBQ_2.0_test.json"
with open(filepath, mode="r", encoding="utf-8") as input:
  lines = input.readlines()
res_json = json.loads(''.join(lines))

In [6]:
dataset_temp = [x["question_text"] for x in res_json]

In [7]:
dataset_temp[:20]

['Что может вызвать цунами?',
 'Кто написал роман «Хижина дяди Тома»?',
 'Кто автор пьесы «Ромео и Джульетта»?',
 'Как называется столица Румынии?',
 'На каком инструменте играл Джимми Хендрикс?',
 'Какой стране принадлежит остров Таити?',
 'Кто создал Зимний дворец в Санкт-Петербурге?',
 'Кто режиссер фильма "Бриллиантовая рука"?',
 'Какой стране принадлежат Канарские острова?',
 'Какой географический объект называется Лимпопо?',
 'Какое современное государство занимает территорию, на которой находился древний ассирийский город Ниневия?',
 'По какой шкале измеряют силу землетрясений?',
 'В какой стране впервые появилась игра в гольф?',
 'Каким видом спорта увлекался президент РФ В.В.Путин?',
 'Как звали отца Северуса Снейпа?',
 'Как звали сына Драко?',
 'Как звали маму Северуса Снейпа?',
 'Как звали создателя факультета Когтевран (Рейвенкло)?',
 'Как звали дракончика Хагрида?',
 "Кто автор исторического романа 'Княжна Тараканова'?"]

In [10]:
len(dataset_temp)

2330

Тут всего 2330 вопросов и нет никакой информации о похожести. Для решения задачи нам понадобиться ещё один датасет.

Хороший способ скачивать с Google Drive

In [11]:
! pip install --upgrade --no-cache-dir gdown

  Obtaining dependency information for gdown from https://files.pythonhosted.org/packages/54/70/e07c381e6488a77094f04c85c9caf1c8008cdc30778f7019bc52e5285ef0/gdown-5.2.0-py3-none-any.whl.metadata
  Obtaining dependency information for filelock from https://files.pythonhosted.org/packages/b9/f8/feced7779d755758a52d1f6635d990b8d98dc0a29fa568bbe0625f18fdf3/filelock-3.16.1-py3-none-any.whl.metadata
  Obtaining dependency information for PySocks!=1.5.7,>=1.5.6 from https://files.pythonhosted.org/packages/8d/59/b4572118e098ac8e46e399a1dd0f2d85403ce8bbaad9ec79373ed6badaf9/PySocks-1.7.1-py3-none-any.whl.metadata

[notice] A new release of pip is available: 23.2.1 -> 24.3.1
[notice] To update, run: pip install --upgrade pip


In [12]:
!gdown --id 11Q2g2aA-n7fYK95XEB3d6m0CWKSt6rob

/Users/dm/PycharmProjects/information_retrieval_technologies_itmo_2024/.venv/lib/python3.12/site-packages/gdown/__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=11Q2g2aA-n7fYK95XEB3d6m0CWKSt6rob
From (redirected): https://drive.google.com/uc?id=11Q2g2aA-n7fYK95XEB3d6m0CWKSt6rob&confirm=t&uuid=b5381e83-3e95-469b-a8ce-9928c878bee2
To: /Users/dm/PycharmProjects/information_retrieval_technologies_itmo_2024/lab3/src/quora_question_pairs_rus.csv.zip
100%|██████████████████████████████████████| 26.7M/26.7M [00:03<00:00, 6.71MB/s]


In [13]:
!unzip quora_question_pairs_rus.csv.zip

Archive:  quora_question_pairs_rus.csv.zip
  inflating: quora_question_pairs_rus.csv  


Это большой датасет вопросов quora, который был автоматически переведён.

In [14]:
import pandas as pd
quora_df = pd.read_csv('quora_question_pairs_rus.csv', usecols=lambda x: 'Unnamed' not in x)

In [15]:
quora_df.shape

(404288, 3)

In [16]:
quora_df.sample(20)

question1  \
370902        какие преимущества и недостатки носить юбку   
241232  что лучший университет университета в области ...   
319234  было бы ли это формой жестокого обращения с де...   
115469         что помогает исцелить ваше разбитое сердце   
21087                 что могут узнать индейцы из японцев   
11002   какой номер технической поддержки для мобильны...   
111877  как я могу доказать своей девушке, что я не об...   
161795  будущее рождается из прошлого, как вы можете о...   
16394   как бы вы сравнили законы эвтаназии в Швейцари...   
135006  как часто сотрудники должны получать оценку эф...   
154886                         какие ребята любят девушек   
172260  индийское правительство демонизировало 500 и 1...   
99615   как я могу блокировать сайты на моем телефоне ...   
59965   какова цена и разница между ubergo uberx и ube...   
163699  мы можем перевести деньги из sbi cdm на другой...   
10359   как я могу соблазнить прямую девушку, если я л...   
288429           темная материя и темная энергия: энергия   
89322                             делает сок лайма плохим   
149550                                    что такое кусты   
306419  как я взломаю армию каракули 2 мини-милиции, ч...   

                                                question2  is_duplicate  
370902    это макси-юбка, считающаяся официальной одеждой             0  
241232  Каковы лучшие университеты Великобритании по э...             1  
319234  Каково это - проводить много времени в одиноче...             0  
115469             любые способы исцелить разбитое сердце             1  
21087                 что могут узнать индейцы из японцев             1  
11002   который является лучшим номером поддержки хост...             1  
111877    сколько общих расходов вместе с комиссией за ms             0  
161795  что является правильным: я родился или родился...             0  
16394   как бы вы сравнили законы эвтаназии Соединенны...             0  
135006  как часто сотрудники Google получают обзоры пр...             0  
154886                  какой тип девушек нравится парням             0  
172260  запрет на 500 и 1000 рупийских нот оценивается...             1  
99615       как заблокировать youtube на Android-телефоне             0  
59965   который имеет самую высокую разницу цен между ...             0  
163699  это sbi лучший вариант для конвертирования inr...             0  
10359   Я лесбиянка, которая любит прямую девушку, что...             0  
288429                           Что такое темная материя             0  
89322                     как сок лайма становится плохим             1  
149550                  что такое втулка в кустовой муфте             0  
306419  есть ли какие-либо проблемы, зарегистрированны...             0

In [17]:
dataset_temp.extend(quora_df["question1"].values)

In [18]:
dataset_temp.extend(quora_df["question2"].values)

In [19]:
len(dataset_temp)

810906

Скачаем эмбеддинги от Natasha (https://natasha.github.io/navec/)

In [27]:
!pip install navec

!wget https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar


[notice] A new release of pip is available: 23.2.1 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
--2024-11-26 17:37:42--  https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar
Resolving storage.yandexcloud.net (storage.yandexcloud.net)... 2a02:6b8::1d9, 213.180.193.243
Connecting to storage.yandexcloud.net (storage.yandexcloud.net)|2a02:6b8::1d9|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 53012480 (51M) [application/x-tar]
Saving to: ‘navec_hudlit_v1_12B_500K_300d_100q.tar’

navec_hudlit_v1_12B 100%[===================>]  50.56M  5.04MB/s    in 9.2s    

2024-11-26 17:37:51 (5.48 MB/s) - ‘navec_hudlit_v1_12B_500K_300d_100q.tar’ saved [53012480/53012480]



In [28]:
from navec import Navec

path = 'navec_hudlit_v1_12B_500K_300d_100q.tar'
navec = Navec.load(path)

In [29]:
navec.get('python')

array([ 0.19214402, -0.1959362 ,  0.43695402, -0.33588752,  0.39676988,
       -0.07734875,  0.04260164,  0.67186844, -0.04072431, -0.07451263,
        0.49101624,  0.07942662,  0.20798121,  0.18426852, -0.30046153,
       -0.15204567,  0.36486024, -0.85307485,  0.48137093,  0.427026  ,
       -0.6845362 , -0.08942661, -0.3101123 , -0.45371357, -0.611568  ,
       -0.26479366, -0.9259754 , -0.11309931, -0.48673463,  0.42229277,
        0.20572862, -0.13167205, -0.8779028 ,  0.07185546,  0.02660668,
       -0.0868782 ,  0.572537  ,  0.6779026 , -0.11222838,  0.14546263,
       -0.03367202,  0.24061641,  0.08572771,  0.00990655,  0.25097123,
        0.01056161, -0.51034373,  0.09052686,  0.43652472,  0.3318286 ,
       -0.11107466, -0.14714225, -0.34816197, -0.3292252 ,  0.33444104,
       -0.07015625,  0.00249229,  0.03096647, -0.39550957, -0.37711838,
       -0.15093368, -0.3094792 ,  0.17217222,  0.21431988, -0.6753277 ,
       -0.5604233 ,  0.46426848,  0.08588701,  0.44196838,  0.34

In [30]:
import numpy as np

In [31]:
def cosine(u, v):
  res = np.dot(u, v) / (np.linalg.norm(u) * np.linalg.norm(v))
  return res

In [33]:
print(cosine(navec['python'],navec['java']))

0.63120574


In [34]:
print(cosine(navec['python'],navec['змея']))

-0.008952811


In [35]:
print(cosine(navec['питон'],navec['змея']))

0.36701795


In [36]:
print(cosine(navec['питон'],navec['java']))

0.03047316


In [37]:
print(cosine(navec['индонезия'],navec['java']))

0.16179799


In [38]:
print(cosine(navec['индонезия'],navec['ява']))

0.51982415


In [39]:
from nltk.tokenize import WordPunctTokenizer

tokenizer = WordPunctTokenizer()

In [40]:
dataset = dataset_temp.copy()

Строим эмбеддинги для предложений по эмбеддингам для слов

In [41]:
def embed(text):
  res = []

  tokens = tokenizer.tokenize(str(text).lower())

  for token in tokens:
    if token in navec:
      res.append(navec[token])

  if not res:
    res.append(navec['<unk>']) # это из документации

  return np.mean(res, axis=0)

In [42]:
embed("Тестовый текст - это просто текст для теста")

array([-0.09564245, -0.01535456,  0.12325782, -0.06663185, -0.20047495,
        0.03183787,  0.26239553,  0.35224754,  0.2521219 , -0.07307456,
       -0.307853  ,  0.01662676,  0.02816503,  0.06181916,  0.06616659,
       -0.08646315, -0.01595033, -0.18428268,  0.03257051,  0.30827984,
       -0.1424056 , -0.12214033, -0.06059988,  0.34464318,  0.06410901,
       -0.12486495, -0.33976468,  0.38941613, -0.05564322, -0.08336477,
       -0.00159295,  0.30793437,  0.5169112 ,  0.39479265, -0.09011184,
        0.08616064, -0.04342553, -0.01709452,  0.00687668,  0.04990578,
        0.09723997,  0.06334966,  0.0321274 , -0.07932533,  0.23986931,
       -0.08293863, -0.00717402,  0.05017046,  0.05389752, -0.15716644,
       -0.1395224 , -0.25135663, -0.22502461,  0.01293752,  0.05510895,
        0.2168825 , -0.16007565, -0.01202961, -0.2204808 ,  0.16102086,
        0.20584778, -0.39431804, -0.0983623 ,  0.00144196,  0.09941299,
        0.1484852 ,  0.24532807,  0.3359027 ,  0.13392086,  0.05

In [43]:
embedded_data = [embed(x) for x in dataset] # эмбеддинги строятся примерно минуту

## Решение

In [46]:
def get_similar_queries(query, embedded_dataset, original_dataset, top_n=10):
    """
    Ищет наиболее похожие запросы в датасете.
    
    Параметры:
    - query: текстовый запрос (строка), для которого нужно найти похожие запросы.
    - embedded_dataset: список эмбеддингов для запросов из датасета.
    - original_dataset: список текстов запросов из датасета.
    - top_n: количество наиболее похожих запросов, которые нужно вернуть.
    
    Возвращает:
    - Список топ-N похожих запросов из оригинального датасета.
    """
    # Построение эмбеддинга для запроса
    query_embedding = embed(query)
    
    # Вычисление косинусного сходства запроса со всеми эмбеддингами датасета
    similarities = [
        cosine(query_embedding, emb) for emb in embedded_dataset
    ]
    
    # Индексы топ-N наиболее похожих запросов
    top_indices = np.argsort(similarities)[-top_n:][::-1]
    
    # Возврат текстов запросов из оригинального датасета
    return [original_dataset[i] for i in top_indices]


In [47]:
def evaluate_model_custom(user_queries, reference_queries, embedded_data, dataset):
    """
    Тестирование модели на пользовательских запросах.
    
    Параметры:
    - user_queries: список пользовательских запросов для тестирования.
    - reference_queries: список списков эталонных запросов для каждого пользовательского запроса.
    - embedded_data: эмбеддинги всего датасета.
    - dataset: оригинальные данные (вопросы).
    
    Возвращает:
    - Средний коэффициент семантической близости.
    """
    scores = []
    for i, query in enumerate(user_queries):
        # Получение результатов модели для текущего запроса
        model_results = get_similar_queries(query, embedded_data, dataset, top_n=5)
        
        # Эмбеддинги эталонных запросов
        reference_embeddings = [embed(ref) for ref in reference_queries[i]]
        # Эмбеддинги результатов модели
        model_embeddings = [embed(res) for res in model_results]
        
        # Вычисление семантической близости
        similarities = [
            cosine(ref_emb, model_emb)
            for ref_emb in reference_embeddings
            for model_emb in model_embeddings
        ]
        avg_similarity = np.mean(similarities) if similarities else 0.0
        scores.append(avg_similarity)
        
        # Вывод результатов для каждого запроса
        print(f"Запрос: {query}")
        print(f"Модель предложила:")
        for j, result in enumerate(model_results, 1):
            print(f"  {j}. {result}")
        print(f"Средняя семантическая близость: {avg_similarity:.4f}")
        print("-" * 50)
    
    # Общий результат
    overall_score = np.mean(scores)
    print(f"Средний коэффициент семантической близости по всем запросам: {overall_score:.4f}")
    return overall_score


In [48]:
# Пример пользовательских запросов
user_queries = [
    "Как начать программировать?",
    "Лучшие книги по истории",
    "Как приготовить борщ?",
    "Что делать при простуде?",
    "Где купить автомобиль?",
    "Лучшие смартфоны 2024 года",
    "Как улучшить свои навыки общения?",
    "Какие фильмы посмотреть на выходных?",
    "Как найти работу программистом?",
    "Советы для путешествий"
]

# Пример эталонных запросов
reference_queries = [
    ["Как стать программистом?", "С чего начать изучение программирования?", "Как выучить Python?"],
    ["Топ книг по истории", "Какие книги по истории стоит прочитать?", "Книги по всемирной истории"],
    ["Как варить борщ?", "Рецепт борща", "Как приготовить украинский борщ?"],
    ["Как быстро вылечить простуду?", "Что делать при первых признаках простуды?", "Как лечить простуду дома?"],
    ["Где можно купить машину?", "Лучшие сайты для покупки авто", "Как выбрать автомобиль?"],
    ["Рейтинг смартфонов 2024 года", "Лучшие телефоны в этом году", "Какие смартфоны популярны в 2024?"],
    ["Как улучшить общение с людьми?", "Советы по общению", "Как стать лучше в разговоре?"],
    ["Лучшие фильмы для выходных", "Что посмотреть в пятницу?", "Фильмы на Netflix для выходных"],
    ["Как найти работу программистом без опыта?", "Как стать junior-разработчиком?", "Где искать вакансии программисту?"],
    ["Советы для путешественников", "Что нужно знать перед поездкой?", "Как подготовиться к путешествию?"]
]

# Вызов функции
evaluate_model_custom(user_queries, reference_queries, embedded_data, dataset)


Запрос: Как начать программировать?
Модель предложила:
  1. как начать программирование
  2. как начать программирование
  3. как начать биткойн
  4. как начать некоммерческую
  5. как я могу начать изучать информатику и программировать
Средняя семантическая близость: 0.4958
--------------------------------------------------
Запрос: Лучшие книги по истории
Модель предложила:
  1. каковы лучшие книги по истории
  2. каковы лучшие книги по истории
  3. каковы лучшие книги по истории
  4. каковы лучшие книги по истории любви
  5. каковы лучшие книги по истории любви
Средняя семантическая близость: 0.8010
--------------------------------------------------
Запрос: Как приготовить борщ?
Модель предложила:
  1. как приготовить стейк
  2. как приготовить стейк
  3. как правильно приготовить гамбургер
  4. как приготовить гамбургеры
  5. как я могу приготовить холодный суп
Средняя семантическая близость: 0.5740
--------------------------------------------------
Запрос: Что делать при простуде?


0.5882596